[3.x] Add access to the viewport's G-buffers (and depth).

SIsilicon opened the pull request:
With this addition, the G buffers in viewports, and the depth buffer, are now accessible textures that can be used in shaders.

Using the new `buffer_mode` property in `ViewportTexture`, the following buffers are now accessible, including the default `Color` buffer.
- Depth
- Diffuse + Ambient Factor
- Specular + Metallic
- Normal + Roughness
- Subsurface

It is now possible to do various mid-processing effects, such as this.
![temp_annotation](https://user-images.githubusercontent.com/34734122/82576852-d1097100-9b4f-11ea-917a-ac1c23d2d80e.png)

There is also an `expose_gbuffer` variable in `Viewport` that enables access to the G-buffers, with some exceptions.

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/798.*

SIsilicon committed:
Added access to the viewport's other buffers, and a force_mrt variable to Viewport. With this addition, the G buffers in viewports, and the depth buffer, are now accesible textures that can be used in shaders.

Zireael07 commented:
Regarding CI - it failed on clang format and/or documentation, if I am reading it correctly.

SIsilicon commented:
Yeah, the error it's giving is kinda confusing, this is my first time writing in C++ let alone using the Clang. I can't tell what _exactly_ is not right with the formatting in the code parts it's highlighting. :/

SIsilicon committed:
Fixed Clang errors

SIsilicon commented:
Well I'm confused. Can someone explain to me what this is supposed to mean?
![image](https://user-images.githubusercontent.com/34734122/82582013-06658d00-9b57-11ea-87c9-186be1099e06.png)

Calinou commented:
@SIsilicon This is probably due to trailing whitespace. Make sure you don't have trailing whitespace in your local copy.

SIsilicon committed:
Fix clang errors (again)

SIsilicon committed:
fix documentation

SIsilicon committed:
Fix documentation (again)

SIsilicon commented:
Sorry for the commits. I have no idea how to do the check locally myself.
Anyway, there is one more thing I discovered I needed to do. Apparently, Android and iOS failed to build because of usage of texture swizzling. I'll have to have that pre-processed out.

clayjohn commented on drivers/gles3/rasterizer_storage_gles3.cpp:
```diff
@@ -7879,10 +8023,37 @@ bool RasterizerStorageGLES3::free(RID p_rid) {
+		bool buffers_active = rt->buffers.active;
+		bool buffer_effects_active = rt->buffers.effects_active;
 		_render_target_clear(rt);
+		/*RID *rt_textures[6]{ &(rt->texture), &(rt->depth_texture), &(rt->buffers.diffuse_texture),
```
Comments can be removed

clayjohn commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -5147,15 +5193,23 @@ void RasterizerStorageGLES2::render_target_set_size(RID p_render_target, int p_w
+		case VS::VIEWPORT_TEXTURE_BUFFER_DEPTH:
+			return rt->depth_texture;
+		default:
+			//printf("The buffer ", p_buffer, " is not available in GLES2.");
```
This can be a ``WARN_PRINT()``

clayjohn commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -34,9 +34,12 @@
 #include "servers/camera/camera_feed.h"
 #include "servers/visual/visual_server_raster.h"
 
+#include <iostream>
```
iostream should not be included.

Looks like you just used it while developing, so it should be removed.

clayjohn commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -4515,8 +4520,49 @@ void RasterizerSceneGLES3::render_scene(const Transform &p_cam_transform, const
 
-	if (use_mrt) {
+	// resolve buffers to textures
+	if (storage->frame.current_rt->buffers.active && use_mrt) {
```
This blit should probably only take place when ``force_mrt`` is true. Otherwise these blits will take place every frame for everyone who uses mrts.

clayjohn commented on drivers/gles3/rasterizer_storage_gles3.cpp:
```diff
@@ -7233,6 +7253,30 @@ void RasterizerStorageGLES3::_render_target_allocate(RenderTarget *rt) {
 		glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER, rt->buffers.diffuse);
 
+		GLuint diffuse;
+		glGenTextures(1, &diffuse);
```
This goes the same for the others below:

You should only be allocating space for textures when forcing mrts. Otherwise you are doubling the memory cost for all users for a feature that they likely aren't using.

SIsilicon commented:
Ok, thanks for pointing all that out @clayjohn . I guess I was a little messy with my code. 😅
As for the wasted memory thing, I guess I should take your suggestions, plus I'll rename `force_mrt` to `expose_gbuffer` since that name makes more sense with the proposed changes.

clayjohn reviewed (changes requested):
Looking good so far!

As you'll see from my comments, I am concerned about the constant memory/performance impact this PR would have. However, I think any performance issues can be addressed with a slight API change. 

Right now the extra textures are created regardless of whether the user is using mrts, and then if mrts are used the blits take place automatically, even if the textures are not read. Even if a user does not use mrts, they can force mrts in order to copy the textures. The problem there is that they are then forced to calculate mrts even if there scene doesn't benefit from them.

I suggest instead exposing a boolean ``expose_textures``, the textures would only be exposed when ``expose_textures`` is true. Further, you would only expose the textures that would be used (so mrts will only be exposed when mrts are used). Your block in ``render_scene`` would become (in psuedo-code):
```
if (expose_textures) {
    //blit color
    //blit depth
    if (use_mrts) {
        //blit mrts
    }
}
```

Note: this is just my view of a better API, making the requested changes doesn't guarantee that this will be merged. The final decision is up to reduz and not me.

SIsilicon commented:
Technically, color is already exposed, and depth already had its own texture, so they wouldn't need to be explicitly exposed. This would make more sense for the gbuffers however, as as you said before, their textures may not always be accessed. So how about this?
```
// Scene storage
func set_expose_gbuffer()  {
    Clear render target
    Update render target depending on expose_gbuffer.
    // This includes creating the textures.

...

// Scene render
if (expose_gbuffer) {
    //blit gbuffers
}
```

clayjohn commented:
Yea, that sounds good. I think lots of people are familiar with the term "g buffers".

The important thing is making sure that this change is zero cost for people who won't use it.

SIsilicon committed:
Optimized texture buffer allocation.
Now the buffers will only be created when `expose_gbuffer` is enabled in a Viewport.
Plus texture swizzling on the depth buffer will no longer attempt to happen in WebGL.

SIsilicon commented:
Ok, this should work now (Pardon the late comment). :)

clayjohn commented:
Great work! We just need reduz' approval since this is a fairly big change. 

I will try to get it discussed at the next PR review meeting.

SIsilicon commented:
Just so you guys know, If this pull request goes through, You can count on me to implement this feature in 4.0 as well once the alpha version released. I may not have Vulkan support on my PC, but I can still translate my current code to the best of my ability. And I could even ask for someone's help to debug it. :)

aaronfranke commented:
> @SIsilicon If this pull request goes through, You can count on me to implement this feature in 4.0 as well once the alpha version released. 

Just FYI, this PR won't be merged until after the master branch version is implemented and merged. Generally speaking, pull requests should be made for the master branch first, then changes can be cherry-picked for stable branches. For this change in particular, there seem to be a lot of concerns with adding this feature and personally I wouldn't cherry-pick this for 3.2.

SIsilicon commented:
> @aaronfranke Just FYI, this PR won't be merged until after the master branch version is implemented and merged. Generally speaking, pull requests should be made for the master branch first, then changes can be cherry-picked for stable branches. For this change in particular, there seem to be a lot of concerns with adding this feature and personally I wouldn't cherry-pick this for 3.2.

Understood.

aaronfranke commented:
@SIsilicon There is no purpose in waiting for 4.0 alpha to work on this feature, you can do it anytime.

mindinsomnia commented:
Did this feature make it into Godot 3.2.2?

aaronfranke commented:
@mindinsomnia Do you see how the status says "Open"? This means that it is not in any Godot version yet.

SIsilicon commented:
Unfortunately, I can't seem to get it working for to the lack of Vulkan support. I suppose my contribution will have to wait until I can at least run the engine.

nsrosenqvist commented:
Don't mean to add noise, just wondering if there is anything still blocking this? I need this for a feature and it looks like it just keeps getting pushed back

Calinou commented:
> Don't mean to add noise, just wondering if there is anything still blocking this? I need this for a feature and it looks like it just keeps getting pushed back

This feature needs to be reimplemented on the `master` branch, as we don't merge `3.x` features that don't have an equivalent in the `master` branch. Otherwise, people upgrading from Godot 3.x to Godot 4.0 could end up relying on features that are no longer available in Godot 4.0.

nsrosenqvist commented:
> > Don't mean to add noise, just wondering if there is anything still blocking this? I need this for a feature and it looks like it just keeps getting pushed back
> 
> This feature needs to be reimplemented on the `master` branch, as we don't merge `3.x` features that don't have an equivalent in the `master` branch (unless the feature makes no sense to be supported there). Otherwise, people upgrading from Godot 3.x to Godot 4.0 could end up relying on features that are no longer available in Godot 4.0 because they weren't reimplemented in time. As said above, SIsilicon does not have access to a GPU that runs Vulkan. This means they can't look into implementing this in `master`.
> 
> Even if merging this PR for `3.x` was a goal, it would need to be rebased – there are a lot of conflicts to resolve.

I understand, thanks for clarifying the situation!

nsrosenqvist commented:
> Even if merging this PR for `3.x` was a goal, it would need to be rebased – there are a lot of conflicts to resolve. (To clarify, this feature _is_ desired as long as it doesn't incur a performance penalty for those who don't use it.)

Actually my GPU doesn't support vulkan either, so it was support in the 3.x series I was hoping for 🙂 but it wouldn't be merged if a viable implementation existed for 4.0 too? Anyway one can throw money at this and maybe someone else will take a look at vulkan? Does the Godot project use bounty source or equivalent?

Calinou commented:
> but it wouldn't be merged if a viable implementation existed for 4.0 too?

This would most likely be merged in `3.x`. I can't make any guarantees about it though.

> Anyway one can throw money at this and maybe someone else will take a look at vulkan? Does the Godot project use bounty source or equivalent?

Godot does not use a bounty system, and there are no plans to add one.

bfloch commented:
@SIsilicon This is really amazing work.

Small remarks:
- I noticed that `render_target_v_flip` is not respected.
- I don't seem to get any Normals if a shader has `Transparency` enabled in a spatial shader. Any idea why?

Also: Would there be a way to access the buffers directly in a screen shader, without having to create ViewportTexture uniforms?

Zireael07 commented:
Just mentioning that this PR would make effects such as https://godotengine.org/qa/123831/how-to-do-outline-and-overlay-of-3d-meshes possible (or significantly easier and more performant)

GloomsfieldRooster commented:
hello!! any updates on this?

Calinou commented:
> hello!! any updates on this? (for godot 4 specifically)

This PR needs a reimplementation from scratch for Godot 4, but SIsilicon no longer seems to be available to do this. Also, a reimplementation would need significant design changes to follow what's being done in https://github.com/godotengine/godot/pull/80214.

GloomsfieldRooster commented:
> This PR needs a reimplementation from scratch for Godot 4, but SIsilicon no longer seems to be available to do this. Also, a reimplementation would need significant design changes to follow what's being done in #80214.

i see, thank you!!

aaronfranke commented:
Is this feature still needed now that Godot 4.5 has stencil buffer support?

SIsilicon commented:
It's been a while hasn't it. I've left the Godot community years ago, but it wouldn't feel right to leave this project unfinished. I don't have as much free time as I used to, but if someone can fill me in on any changes that need to be made and what happened in Godot 3.x, I would be very grateful.
